src/workers/newView.ts

```diff
@@ -24,12 +24,78 @@ const addView = async (con: DataSource, entity: View): Promise<boolean> => {
   return false;
 };
 
+const DEFAULT_TIMEZONE = 'UTC'; // in case user doesn't have a timezone set
+const INC_STREAK_QUERY = `
+WITH u AS (
+  SELECT "id",
+      /* Increment the current streak if
+        a) lastViewAt is NULL - this is a new user, we should start a new streak
+        b) the currentStreak is 0 - we should start a new streak
+        c) we didn't do it today already
+      */
+      ("lastViewAt" is NULL OR
+       "currentStreak" = 0 OR
+       DATE("lastViewAt" AT TIME ZONE COALESCE("timezone", $3)) <> DATE($2 AT TIME ZONE COALESCE("timezone", $3))
+      ) AS shouldIncrementStreak,
+
+      -- Increment maxStreak if it's the same as currentStreak
+      "currentStreak" = "maxStreak" AS shouldUpdateMaxStreak
+  FROM public.user AS users
+  INNER JOIN user_streak ON user_streak."userId" = users.id
+  WHERE users.id = $1
+)
+UPDATE user_streak AS us SET
+  "lastViewAt" = $2,
+  "updatedAt" = now(),
+  "currentStreak" = CASE WHEN shouldIncrementStreak THEN us."currentStreak" + 1 ELSE us."currentStreak" END,
+  "totalStreak" = CASE WHEN shouldIncrementStreak THEN us."totalStreak" + 1 ELSE us."totalStreak" END,
+  "maxStreak" = CASE WHEN shouldIncrementStreak AND shouldUpdateMaxStreak THEN us."maxStreak" + 1 ELSE us."maxStreak" END
+FROM u
+WHERE us."userId" = u.id
+RETURNING shouldIncrementStreak
+`;
+
+const incrementReadingStreak = async (
+  con: DataSource,
+  data: DeepPartial<View>,
+): Promise<boolean> => {
+  const repo = con.getRepository(UserStreak);
+  const { userId, timestamp } = data;
+
+  if (!userId) {
+    return false;
+  }
+
+  const viewTime = timestamp ? new Date(timestamp as string) : new Date();
```

Comment:
So in this case, the incoming timestamp is stored in the db with the timezone the database is in, correct? Thus we are casting it to the proper timezone in the query above

Reply:
I tested locally with nodejs runtime timezone set to PST and it all works as it's supposed to: https://dailydotdev.slack.com/archives/C05TQL6HWTH/p1706631354151889?thread_ts=1706625076.569029&cid=C05TQL6HWTH